lib/virsh_autotest_base.pm

```diff
@@ -0,0 +1,126 @@
+# VIRSH TEST MODULE BASE PACKAGE
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+#
+# Summary: This is the base package for virsh test modules, for example,
+# tests/virtualization/xen/hotplugging.pm
+# tests/virt_autotest/virsh_internal_snapshot.pm
+# tests/virt_autotest/virsh_external_snapshot.pm
+# and etc.
+# Maintainer: Wayne Chen <wchen@suse.com>
+
+package virsh_autotest_base;
+
+use base "consoletest";
+use strict;
+use warnings;
+use POSIX 'strftime';
+use File::Basename;
+use Data::Dumper;
+use XML::Writer;
+use IO::File;
+use testapi;
+use utils;
+use virt_utils;
+use xen;
+
+sub run_test {
+    die('Please override this subroutine in children modules to run desired tests.');
+}
+
+sub run {
+    my ($self) = @_;
+    $self->{"start_run"} = time();
+    $self->run_test;
+    $self->{"stop_run"} = time();
+    return unless get_var("VIRT_AUTOTEST");
+    $self->junit_log_provision((caller(0))[3]);
+}
+
+sub junit_log_provision {
+    my ($self, $runsub) = @_;
+    my $status    = eval { $runsub =~ /post_fail_hook/img ? 'FAILED' : 'PASSED' };
+    my $tc_result = $self->analyzeResult($status);
+    $self->junit_log_params_provision($tc_result);
+    ###Load instance attributes into %stats
+    my %stats;
+    foreach (keys %{$self}) {
+        if (defined($self->{$_})) {
+            if (ref($self->{$_}) eq 'HASH') {
+                %{$stats{$_}} = %{$self->{$_}};
+            }
+            elsif (ref($self->{$_}) eq 'ARRAY') {
+                @{$stats{$_}} = @{$self->{$_}};
+            }
+            else {
+                $stats{$_} = $self->{$_};
+            }
+        }
+        else {
+            next;
+        }
+    }
+    print "The data to be used for xml generation:", Dumper(\%stats);
+    my $xml_result = generateXML_from_data($tc_result, \%stats);
+    script_run "echo \'$xml_result\' > /tmp/output.xml";
+    save_screenshot;
+    parse_junit_log("/tmp/output.xml");
+}
+
+sub junit_log_params_provision {
+    my ($self, $data) = @_;
+    my %my_hash   = %$data;
```

Comment:
you calling `junit_log_params_provision` only after `analyzeResult` function which return `$result` .  what if : 
1. you will call `junit_log_params_provision` **inside**  `analyzeResult` so you won't need to create extra variable just to pass result to `analyzeResult`
2. you will pass it as hash initially so you don't need to do extra conversion

Reply:
This does not need to be changed here. Thanks for your advice anyway.

Reply:
If above is not very convincing to you, any following improvement or enhancement that involves per-guest or per-test-item should be implemented in following actions. Actually, any modification here requires further tuning or tweak in other subroutines or modules. This is beyond the scope of this PR.